### PR TITLE
Generate admin ui according to the flag value

### PR DESCRIPTION
**Pull request by avdev4j**
Add a --with-admin-ui flag to generate the admin ui pages. By default, those pages are not generated and the user should use the JHipster Control Center application.

EDIT: add prompt instead

#12167 
#12090

-   [x] Angular
-   [x] React
-   [x] Vue
-   [x] i18n
-   [x] JDL

---

Please make sure the below checklist is followed for Pull Requests.

-   [x] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary https://github.com/jhipster/jhipster-online/pull/257
-   [X] Documentation is added/updated where necessary https://github.com/jhipster/jhipster.github.io/pull/1026
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mraible** [comment]
Does this mean we're going to require two apps to run one app (if you want to administer the one app)?

**Comment by avdev4j** [comment]
> Does this mean we're going to require two apps to run one app (if you want to administer the one app)?

Unless we use the flag I have added ;).

File: generators/client/index.js [hunk 1]
```diff
@@ -63,6 +63,12 @@ module.exports = class JHipsterClientGenerator extends BaseBlueprintGenerator {
             type: Boolean,
         });
 
+        // This adds support for a `--with-admin-ui` flag which can be used to generate the admin UI pages
+        this.option('with-admin-ui', {
+            desc: 'Generate the UI admin pages. By default thi feature is supported by the JHipster Control Center.',
```
Comment (MarlonLuan) [inline generators/client/index.js:68]:
```suggestion
            desc: 'Generate the UI admin pages. By default this feature is supported by the JHipster Control Center.',
```

**Review by MarlonLuan — COMMENTED** [review]
Typo

**Comment by pascalgrimaud** [comment]
I'm not totally in favor of this, for now. 
I'd prefer to have no-admin-ui flag instead.

The reason is the JHipster Control Center is not finished yet today, and we don't have official lead on this. So it can become a dead project, if it's not well maintained.

**Comment by avdev4j** [comment]
> I'm not totally in favor of this, for now.
> I'd prefer to have no-admin-ui flag instead.

This was my main issue yesterday, Maybe we should add a question, something like (Would like to administrate your application with the JHipster Control Center?(the admin ui will be not generated) y/N".

Because if we add a flag like no-admin-ui I'm sure that nobody will use it :D?

With this question we will be able to add the JHCC Docker compose configuration only when the answer is "yes". 

WDYT?

**Comment by PierreBesson** [comment]
I would say let's generate the jhcc docker compose all the time so people can try it out in any case.... But I'm conflicted with Pascal's idea that maybe the admin pages should stay as default. As for if the admin UI should be generated by default or not, I think it would be better to actually do a vote about it on the mailing list.

File: generators/generator-base.js [hunk 2]
```diff
@@ -2325,6 +2328,7 @@ module.exports = class JHipsterBaseGenerator extends PrivateBase {
         dest.experimental = config.experimental;
         dest.logo = config.logo;
         dest.backendName = config.backendName || 'Java';
+        dest.withAdminUi = config.withAdminUi;
```
Comment (mshima) [inline generators/generator-base.js:2331]: Move to loadAppConfig.

File: generators/generator-base.js [hunk 3]
```diff
@@ -2238,6 +2238,9 @@ module.exports = class JHipsterBaseGenerator extends PrivateBase {
         if (options.recreateInitialChangelog) {
             this.configOptions.recreateInitialChangelog = options.recreateInitialChangelog;
         }
+        if (options.withAdminUi) {
+            this.configOptions.withAdminUi = options.withAdminUi;
```
Comment (mshima) [inline generators/generator-base.js:2242]:
```suggestion
        if (options.withAdminUi !== undefined) {
            this.jhipsterConfig.withAdminUi = options.withAdminUi;
```
For reproducibility.

File: generators/client/index.js (hunk 1 above)
Comment (avdev4j) [reply generators/client/index.js]: Done!

File: generators/generator-base.js (hunk 3 above)
Comment (avdev4j) [reply generators/generator-base.js:2242]: Done, what is the difference between configOptions and jhipsterConfig?

**Comment by pascalgrimaud** [comment]
It needs to be discussed, when admin ui is not choosen, should we keep swagger or not ?
I'd say:
- with admin ui: like today
- without admin ui: user-management + swagger

Swagger is so important, and it would not required too much maintenance !

**Comment by avdev4j** [comment]
after all, I think you're right @pascalgrimaud it's not a big effort to keep that so let's revert this changes.

**Comment by avdev4j** [comment]
@MathieuAA I will maybe ping you to know how we can introduce that in the JDL Application (and if it's a good idea).

**Comment by MathieuAA** [comment]
@avdev4j It's an application option, right? So if it's a matter of adding another option, it's no big deal.

File: test/jdl/jhipster/default-application-options.spec.js [hunk 4]
```diff
@@ -60,6 +60,9 @@ describe('DefaultApplicationOptions', () => {
             it('should set the client theme variant option to none', () => {
                 expect(options.clientThemeVariant).to.equal('');
             });
+            it('should set withAdminUI option to true', () => {
+                expect(options.withAdminUi).to.equal(true);
```
Comment (MathieuAA) [inline test/jdl/jhipster/default-application-options.spec.js:64]: A small improvement: `expect(options.withAdminUi).to.equal(true);` => `expect(options.withAdminUi).to.be.true;`

**Review by MathieuAA — CHANGES_REQUESTED** [review]
I've quickly reviewed the JDL part.
I'll do a more thorough review for this.

File: generators/client/templates/angular/src/main/webapp/app/layouts/navbar/navbar.component.html.ejs [hunk 5]
```diff
@@ -80,6 +80,7 @@
                         </a>
                     </li>
                     <%_ } _%>
+                    <%_ if(withAdminUi){ _%>
```
Comment (MathieuAA) [inline generators/client/templates/angular/src/main/webapp/app/layouts/navbar/navbar.component.html.ejs:83]:
```suggestion
                    <%_ if (withAdminUi) { _%>
```

File: generators/client/prompts.js [hunk 6]
```diff
@@ -188,3 +189,23 @@ function askForClientThemeVariant() {
...
 }
+
+function askForAdminUi() {
+    if (this.existingProject) {
+        return undefined;
+    }
+
+    const skipClient = this.skipClient;
+
+    const PROMPT = {
```
Comment (MathieuAA) [inline generators/client/prompts.js:200]: why is this uppercase?

File: generators/client/templates/react/src/main/webapp/app/modules/administration/administration.reducer.ts.ejs [hunk 7]
```diff
@@ -71,31 +75,39 @@ export default (state: AdministrationState = initialState, action): Administrati
...
     <%_ } _%>
+  <%_ if (withAdminUi) { _%>
     case REQUEST(ACTION_TYPES.FETCH_METRICS):
     case REQUEST(ACTION_TYPES.FETCH_THREAD_DUMP):
     case REQUEST(ACTION_TYPES.FETCH_LOGS):
     case REQUEST(ACTION_TYPES.FETCH_CONFIGURATIONS):
     case REQUEST(ACTION_TYPES.FETCH_ENV):
     case REQUEST(ACTION_TYPES.FETCH_HEALTH):
+  <%_ } _%>
+  <%_ if (withAdminUi ||  (applicationType === 'gateway' && serviceDiscoveryType)) { _%>
```
Comment (MathieuAA) [inline generators/client/templates/react/src/main/webapp/app/modules/administration/administration.reducer.ts.ejs:86]:
```suggestion
  <%_ if (withAdminUi || (applicationType === 'gateway' && serviceDiscoveryType)) { _%>
```

File: generators/client/templates/react/src/main/webapp/app/modules/administration/administration.reducer.ts.ejs [hunk 8]
```diff
@@ -71,31 +75,39 @@ export default (state: AdministrationState = initialState, action): Administrati
...
     <%_ } _%>
+  <%_ if (withAdminUi) { _%>
     case FAILURE(ACTION_TYPES.FETCH_METRICS):
     case FAILURE(ACTION_TYPES.FETCH_THREAD_DUMP):
     case FAILURE(ACTION_TYPES.FETCH_LOGS):
     case FAILURE(ACTION_TYPES.FETCH_CONFIGURATIONS):
     case FAILURE(ACTION_TYPES.FETCH_ENV):
     case FAILURE(ACTION_TYPES.FETCH_HEALTH):
+  <%_ } _%>
+  <%_ if (withAdminUi ||  (applicationType === 'gateway' && serviceDiscoveryType)) { _%>
```
Comment (MathieuAA) [inline generators/client/templates/react/src/main/webapp/app/modules/administration/administration.reducer.ts.ejs:104]:
```suggestion
  <%_ if (withAdminUi || (applicationType === 'gateway' && serviceDiscoveryType)) { _%>
```

File: test/jdl/grammar/validator.spec.js [hunk 9]
```diff
@@ -503,6 +503,33 @@ describe('JDLSyntaxValidatorVisitor', () => {
...
+            context('an invalid value', () => {
+                it('should report a syntax error', () => {
+                    expect(() =>
+                        parse(`
+            application {
+              config {
+                withAdminUi 666
+              }
+            }`)
+                    ).to.throw(new RegExp('^A boolean literal is expected, but found: "666"'));
```
Comment (MathieuAA) [inline test/jdl/grammar/validator.spec.js:529]: Wouldn't the following work?

```suggestion
                    ).to.throw(/^A boolean literal is expected, but found: "666"/);
```

File: generators/client/prompts.js (hunk 6 above)
Comment (avdev4j) [reply generators/client/prompts.js:200]: I copied this from another prompt in the client prompt file. If you want I cant rename all the PROMPT to prompt.

File: test/jdl/grammar/validator.spec.js (hunk 9 above)
Comment (avdev4j) [reply test/jdl/grammar/validator.spec.js:529]: Indeed we should use literal regexp, I do the modification for the whole file

File: generators/client/prompts.js (hunk 6 above)
Comment (avdev4j) [reply generators/client/prompts.js:200]: Ok so I think this have been done to avoir misunderstanding between PROMPT and prompt in the callback ;)

Comment (MathieuAA) [reply generators/client/prompts.js:200]: Really? Wow. Talk about dev laziness... I know naming things is one of the most difficult things we devs have to do but still...

Comment (MathieuAA) [reply generators/client/prompts.js:200]: The PROMPT thing is a param object, its name could be "promptParams". The prompt function param is regular param, which represents a prompt result, an answer.

**Comment by avdev4j** [comment]
@MathieuAA I have tested many things and finally did not find the better way, except if we move to async functions.

Less code, easier to read and ofc less "PROMPT" ;).

I have added a commit, about that, in this PR but we can split into two different ones. 

This is related to this PR: https://github.com/jhipster/generator-jhipster/pull/12812

What do you think @mshima @sendilkumarn ?

**Review by mshima — APPROVED** [review]
LGTM.

2 non-blocker considerations:
- I would prefer --admin-ui instead of --with-admin-ui.
- httpsGet should be extracted into a util function instead of a generator-private member.

File: generators/client/prompts.js [hunk 10]
```diff
@@ -113,59 +112,28 @@ function askForClientTheme() {
         { value: 'yeti', name: 'Yeti' },
     ];
 
-    const PROMPT = {
+    const bootSwatchChoices = await retrieveBootswatchThemes(self).catch(() =>
```
Comment (MathieuAA) [inline generators/client/prompts.js:115]: Is it bootSwatch or bootswatch? I'm confused...

**Comment by MathieuAA** [comment]
That's a much better solution @avdev4j :)

**Review by MathieuAA — APPROVED** [review]

File: generators/client/prompts.js (hunk 10 above)
Comment (avdev4j) [reply generators/client/prompts.js:115]: bootswatch indeed ;), let me fix that ;)

**Comment by avdev4j** [comment]
> * I would prefer --admin-ui instead of --with-admin-ui.
 --with-admin-ui no longer exist because we move to a prompt instead of a flag.

> * httpsGet should be extracted into a util function instead of a generator-private member. Not the scope of this PR.
Not scope in this PR but let rework the function to easily extract it in the future.

**Comment by pascalgrimaud** [comment]
congrats for your hard work @avdev4j !

**Comment by pascalgrimaud** [comment]
lot of work done to achieve this, so increasing the bounty, it's well deserved

**Comment by avdev4j** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/27873

**Comment by pascalgrimaud** [comment]
@avdev4j : approved

**Pull request merged**